README.md
# selection

Radio button with focus. Developed for flutter for web but can also be used on Android. iOS is unverified, but probably works..

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.

Here is a new type radio. It is able to select exclusively.
I'll describe how to use.

## How it works

Focus widget is a top-level child.
The radio button for the web doesn't accept the tab key, so I made it.
Please make the content with SelectionManifest<T> class.
Also, define the selection value in enum and supply its name in the <T> part.

# Usage

To use this plugin, add `selection` as a [dependency in your pubspec.yaml file.](https://flutter.dev/docs/development/packages-and-plugins/using-packages)

# Example

```dart
import 'package:flutter/material.dart';
import 'package:selection/selection.dart';

/// For Selection parts begin.
enum Sex {
  notKnown,
  male,
  female,
  not_applicable,
}
List<SelectionManifest<Sex>> _sexManifest = [
  SelectionManifest<Sex>(
      label: 'not known', value: Sex.notKnown, width: 120, height: 40),
  SelectionManifest<Sex>(
      label: 'male', value: Sex.male, width: 90, height: 40),
  SelectionManifest<Sex>(
      label: 'female', value: Sex.female, width: 100, height: 40),
  SelectionManifest<Sex>(
      label: 'not applicable', value: Sex.not_applicable, width: 150, height: 40),
];

enum DominantHand {
  notKnown,
  left,
  right,
}
List<SelectionManifest<DominantHand>> _dominantHandManifest = [
  SelectionManifest<DominantHand>(
      label: 'not known', value: DominantHand.notKnown, width: 120, height: 40),
  SelectionManifest<DominantHand>(
      label: 'left', value: DominantHand.left, width: 80, height: 40),
  SelectionManifest<DominantHand>(
      label: 'right', value: DominantHand.right, width: 90, height: 40),
];
/// For Selection parts end.

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  /// For Selection parts begin.
  final focusNodes = [
    FocusNode(),
    FocusNode(),
  ];
  final _sex = <Sex>[null];
  final _dominant = <DominantHand>[null];
  /// For Selection parts end.

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Center(
            child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  Selection(
                      focusNodes: focusNodes,
                      focusNodeIndex: 0,
                      manifests: _sexManifest,
                      selectedValue: _sex),
                  SizedBox(height: 8),
                  Selection(
                      focusNodes: focusNodes,
                      focusNodeIndex: 1,
                      manifests: _dominantHandManifest,
                      selectedValue: _dominant)
                ])));
  }
}
```

>Selection(
>      {this.key,
>        this.focusNodes,
>        this.focusNodeIndex,
>        this.manifests,
>        this.selectedValue})
>      : super(key: key);

'key' is not used as this class.
'focusNodes' is an array of FocusNodes. It will take the same number of widgets to tabbing.
'focusNodeIndex' is a so-called tab index.
'manifests' describes the nature of the content.
'selectedValue' is designed to use the 0th array so that the value can be changed from the outside.
Type T must be created with enum.

You may need a script to disable tab key in index.html for web.

The color and shape are adjusted to suit my convenience.
I'm sorry for explaining the behavior, please touch example.

Thank you.

CHANGELOG.md
## [0.0.1] - 2019-11-05

Initial release.
It's cold?. It's cold.

## [1.0.0] - 2019-11-05
I made it.

## [1.0.0+1] - 2019-11-07
Editing descriptions.

## [1.0.1] - 2019-11-07
Fixing web release mode.

## [1.0.1+1] - 2019-11-07
Coordinated homepage address issues.

## [1.0.2] - 2019-11-07
Resolved an exception has occurred in the type cast test.

## [1.0.2+1] - 2019-11-09
Homepage url was fixed on pubspec.yaml.

test/selection_test.dart
//import 'package:flutter_test/flutter_test.dart';

//import 'package:selection/selection.dart';

void main() {
  //test('adds one to input values', () {
  //  final calculator = Calculator();
  //  expect(calculator.addOne(2), 3);
  //  expect(calculator.addOne(-7), -6);
  //  expect(calculator.addOne(0), 1);
  //  expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  //});
}


lib/selection.dart
library selection;

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class SelectionManifest<T> {
  SelectionManifest({this.label, this.value, this.width, this.height});
  @required
  final String label;
  @required
  final T value;
  @required
  final double width;
  @required
  final double height;
}

class Selection<T> extends StatefulWidget {
  Selection(
      {this.key,
      this.focusNodes,
      this.focusNodeIndex,
      this.manifests,
      this.selectedValue})
      : super(key: key);
  static const String className = "Selection";

  final Key key;
  final List<FocusNode> focusNodes;
  final int focusNodeIndex;
  @required
  final List<SelectionManifest<T>> manifests;
  @required
  final List<T> selectedValue;

  @override
  _SelectionState createState() => _SelectionState();
}

class _SelectionState extends State<Selection> {
  bool _alreadyDisposed = false;
  int _focusedIndex;
  bool _focused = false;

  @override
  void initState() {
    super.initState();
    _focusedIndex = widget.selectedValue[0]?.index ?? 0;
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
  }

  @override
  void setState(Function fn) {
    if (_alreadyDisposed) return;

    super.setState(fn);
  }

  @override
  void dispose() {
    _alreadyDisposed = true;
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Focus(
        focusNode: widget.focusNodes[widget.focusNodeIndex],
        onFocusChange: (value) => setState(() => _focused = value),
        child: Wrap(
            children:
                List<Widget>.generate(widget.manifests.length, (int index) {
          return Container(
              padding: EdgeInsets.fromLTRB(0, 0, 8, 0),
              width: widget.manifests[index].width,
              height: widget.manifests[index].height,
              child: Material(
                  borderRadius: BorderRadius.circular(8.0),
                  clipBehavior: Clip.antiAlias,
                  color: _focused && (_focusedIndex ?? -1) == index
                      ? Colors.blue[100]
                      : Colors.grey[300],
                  child: Stack(children: <Widget>[
                    Container(
                        height: widget.manifests[index].height,
                        padding:
                            EdgeInsets.symmetric(horizontal: 8, vertical: 0),
                        child: Row(
                            mainAxisAlignment: MainAxisAlignment.spaceBetween,
                            crossAxisAlignment: CrossAxisAlignment.center,
                            children: <Widget>[
                              Icon(
                                  widget.selectedValue[0]?.index == index
                                      ? Icons.radio_button_checked
                                      : Icons.radio_button_unchecked,
                                  color:
                                      _focused && (_focusedIndex ?? -1) == index
                                          ? Colors.blue
                                          : Colors.black),
                              Text('${widget.manifests[index].label}')
                            ])),
                    InkWell(
                        child: null,
                        onTap: () {
                          FocusScope.of(context).requestFocus(
                              widget.focusNodes[widget.focusNodeIndex]);
                          setState(() {
                            _focused = true;
                            _focusedIndex = index;
                            widget.selectedValue[0] =
                                widget.manifests[index].value;
                          });
                        })
                  ])));
        }).toList()),
        onKey: _onKey);
  }

  bool _onKey(FocusNode node, RawKeyEvent keyEvent) {
    if (node.hasFocus) {
      if (keyEvent is RawKeyUpEvent) {
        /// move focus
        if (keyEvent.data.physicalKey == PhysicalKeyboardKey.tab) {
          if (!keyEvent.data.isShiftPressed) {
            /// next
            int next = widget.focusNodeIndex + 1;
            if (next == widget.focusNodes.length) next = 0;
            FocusScope.of(context).requestFocus(widget.focusNodes[next]);
            return true;
          } else {
            /// prev
            int prev = widget.focusNodeIndex - 1;
            if (prev < 0) prev = widget.focusNodes.length - 1;
            FocusScope.of(context).requestFocus(widget.focusNodes[prev]);
            return true;
          }
        } else if ((keyEvent.data.physicalKey ==
                PhysicalKeyboardKey.arrowDown) ||
            (keyEvent.data.physicalKey == PhysicalKeyboardKey.arrowRight)) {
          /// switch next
          setState(() {
            bool alreadyOn = widget.selectedValue[0] ==
                widget.manifests[_focusedIndex].value;
            final qty = widget.manifests.length;
            _focusedIndex++;
            if (_focusedIndex == qty) _focusedIndex = 0;
            if (alreadyOn)
              widget.selectedValue[0] = widget.manifests[_focusedIndex].value;
          });
          return true;
        } else if ((keyEvent.data.physicalKey == PhysicalKeyboardKey.arrowUp) ||
            (keyEvent.data.physicalKey == PhysicalKeyboardKey.arrowLeft)) {
          /// switch prev
          setState(() {
            bool alreadyOn = widget.selectedValue[0] ==
                widget.manifests[_focusedIndex].value;
            final qty = widget.manifests.length;
            _focusedIndex--;
            if (_focusedIndex < 0) _focusedIndex = qty - 1;
            if (alreadyOn)
              widget.selectedValue[0] = widget.manifests[_focusedIndex].value;
          });
          return true;
        } else if (keyEvent.data.physicalKey == PhysicalKeyboardKey.space) {
          if (widget.selectedValue[0] != widget.manifests[_focusedIndex].value)
            setState(() => widget.selectedValue[0] =
                widget.manifests[_focusedIndex].value);
          return true;
        }
      }
    }
    return false;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:selection/selection.dart';

/// For Selection parts begin.
enum Sex {
  notKnown,
  male,
  female,
  not_applicable,
}
List<SelectionManifest<Sex>> _sexManifest = [
  SelectionManifest<Sex>(
      label: 'not known', value: Sex.notKnown, width: 120, height: 40),
  SelectionManifest<Sex>(label: 'male', value: Sex.male, width: 90, height: 40),
  SelectionManifest<Sex>(
      label: 'female', value: Sex.female, width: 100, height: 40),
  SelectionManifest<Sex>(
      label: 'not applicable',
      value: Sex.not_applicable,
      width: 150,
      height: 40),
];

enum DominantHand {
  notKnown,
  left,
  right,
}
List<SelectionManifest<DominantHand>> _dominantHandManifest = [
  SelectionManifest<DominantHand>(
      label: 'not known', value: DominantHand.notKnown, width: 120, height: 40),
  SelectionManifest<DominantHand>(
      label: 'left', value: DominantHand.left, width: 80, height: 40),
  SelectionManifest<DominantHand>(
      label: 'right', value: DominantHand.right, width: 90, height: 40),
];

/// For Selection parts end.

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  /// For Selection parts begin.
  final focusNodes = [
    FocusNode(),
    FocusNode(),
  ];
  final _sex = <Sex>[null];
  final _dominant = <DominantHand>[null];

  /// For Selection parts end.

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Center(
            child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
              Selection(
                  focusNodes: focusNodes,
                  focusNodeIndex: 0,
                  manifests: _sexManifest,
                  selectedValue: _sex),
              SizedBox(height: 8),
              Selection(
                  focusNodes: focusNodes,
                  focusNodeIndex: 1,
                  manifests: _dominantHandManifest,
                  selectedValue: _dominant)
            ])));
  }
}
